torch/csrc/jit/fusion_compiler.cpp

```diff
@@ -1047,67 +1140,409 @@ struct CPUFusionFunction : public CompiledFusionFunction {
-  auto & graph = *fusion_group->g(attr::Subgraph);
-  AnnotatedGraph agraph(graph, fusion_group->i(attr::device));
-  for(auto & input : graph.inputs()) {
-    auto t = input->type()->expect<TensorType>();
-    agraph.input_desc.emplace_back(t);
+// See Note [Run-time shape checking code] for more explanation on the algorithm.
+at::optional<std::vector<int64_t>> FusedKernelCache::canRunKernel(at::TensorList args) {
+  AT_CHECK(args.size() == input_chunks.size(),
+           "Expected ", input_chunks.size(), " arguments, but got ", args.size());
+
+  at::optional<std::vector<int64_t>> map_size;
+  for (const auto & broadcast_group : input_broadcast_groups) {
+    if (!map_size) {
+      map_size = getMapSize(args, broadcast_group);
+      if (!map_size) {
+        return at::nullopt;
+      }
+    } else {
+      auto group_map_size = getMapSize(args, broadcast_group);
+      // NB: this checks that group_map_size is defined AND equal to map_size
+      if (map_size != group_map_size) {
+        return at::nullopt;
+      }
+    }
   }
-  for(auto & output : graph.outputs()) {
-    auto t = output->type()->expect<TensorType>();
-    agraph.output_desc.emplace_back(t);
+  return map_size;
+}
+
+void FusedKernelCache::runFallback(Stack& stack) {
+  InterpreterState(fallback_code).runOneStage(stack);
+}
+
+void FusedKernelCache::expandArgs(std::vector<at::Tensor>& args, std::vector<int64_t>& map_size) {
+  for (size_t i = 0; i < args.size(); ++i) {
+    auto & arg = args[i];
+    auto & pdesc = input_chunks[i];
+    if (pdesc.nSubtensors == 1) {
+      if (arg.sizes().equals(map_size)) continue;
+      arg = arg.expand(map_size);
+    } else {
+      map_size.at(pdesc.dim) *= pdesc.nSubtensors;
+      if (!arg.sizes().equals(map_size)) {
+        arg = arg.expand(map_size);
+      }
+      map_size.at(pdesc.dim) /= pdesc.nSubtensors;
+    }
   }
-  return getOrCompile(agraph);
 }
 
+std::unique_ptr<FusedKernel> FusedKernelCache::compileSpec(
+      const FusedKernelArgSpec& spec, const std::vector<int64_t>& map_size) {
+  AnnotatedGraph agraph {*graph, device};
 
-std::shared_ptr<CompiledFusionFunction> FusionCompiler::getOrCompile(Graph & graph,
-                                                     int device,
-                                                     at::ArrayRef<at::Tensor> inputs,
-                                                     at::ArrayRef<at::Tensor> outputs) {
-  AnnotatedGraph agraph(graph, device);
-  for(auto & i : inputs) {
-   agraph.input_desc.emplace_back(i);
+  agraph.input_desc = spec.descs();
+  at::optional<at::ScalarType> scalar_type;
+  for (TensorDesc& desc : agraph.input_desc) {
+    if (isFloatingType(desc.scalar_type)) {
+      scalar_type = desc.scalar_type;
+      break;
+    }
   }
-  for(auto & i : outputs) {
-   agraph.output_desc.emplace_back(i);
+  JIT_ASSERT(scalar_type);
+
+  for (Value * output : graph->outputs()) {
```

Comment:
What's going on here?

Reply:
Not sure what exactly do you have in mind? It's pretty much the same code it used to be, except we can no longer use the fact that graph outputs have complete shapes filled in by shape prop.

Reply:
I guess, I don't remember having to adjust the concat sizes in the old code. Why does this adjustment happen here?

Reply:
Because previously the adjustment would be performed by the shape prop? Now, we only know the map_size, and the concat output size is easily derivable from that.